### PR TITLE
Sema: Correct composition of property wrappers.

### PR DESCRIPTION
**Pull request by jckarter**
Fixes rdar://problem/53407949 | SR-11138. Previously, we'd only look at the outermost property wrapper to decide whether the wrapped property's getter and setter were `mutating` (or exist at all). In reality, this requires considering the semantics of the composed accesses of each wrapper layer's `wrappedValue` property. Fixing this systematically addresses a number of issues:
    
- As SR-11138 reported, composing a nonmutating-get-set wrapper ought to produce a composed wrapper that's nonmutating.
    
- We would previously allow a property wrapper with a mutating getter to be nested inside one with only a getter, even though the resulting implementation was unsound (because there's no mutable context for the inner wrapper to execute its get on.)
    
- Similarly, we would construct unsound setters in cases where the setter can't exist, such as when the nested wrapper isn't settable but the outer wrapper is.

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

File: include/swift/AST/PropertyWrappers.h [hunk 1]
```diff
@@ -77,6 +77,47 @@ struct PropertyWrapperTypeInfo {
...
+    // mutating-ness of the outer wrapper's get operation.
+    case Nonmutating:
+      return Getter;
+        
+    // If it's mutating, then it relies
+    // on a) the outer wrapper having a setter to exist at all, and b) the
+    // mutating-ness of either the getter or setter, since we need both to
+    // perform a writeback cycle.
+    case Mutating:
+      if (Setter == DoesntExist) {
```
Comment (brentdax) [inline include/swift/AST/PropertyWrappers.h]: Does this need a special case? If the setter is `DoesntExist`, `std::max()` will always end up returning `DoesntExist`, right?

Comment (jckarter) [reply include/swift/AST/PropertyWrappers.h]: I suppose not. It seems worth calling this out as a specific rule, though.

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Pull request merged**